# Data Engineering 
*Our data needs firstly to be adjusted*. This is the online provided Dataset Card: <br><br>
**Brands**: The brands of smartphones included in the dataset. <br>
**Colors**: The colors available for the smartphones.<br>
**Memory**: The storage capacity of the smartphones, typically measured in gigabytes (GB) or megabytes (MB).<br>
**Storage**: The internal storage capacity of the smartphones, often measured in gigabytes (GB) or megabytes (MB).<br>
**Rating**: The user ratings or scores assigned to the smartphones, reflecting user satisfaction or performance.<br>
**Selling Price**: The price at which the smartphones are sold to consumers.<br>
**Original Price**: The original or list price of the smartphones before any discounts or promotions.<br>
**Mobile**: Indicates whether the device is a mobile phone.<br>
**Discount**: The discount applied to the original price to calculate the selling price.<br>
**Discount percentage**: The percentage discount applied to the original price to calculate the selling price.

### Dataset Adjustment

In [1]:
import pandas as pd 
import numpy as np

In [2]:
df=pd.read_csv('Sales.csv')

In [3]:
df.columns

Index(['Brands', 'Models', 'Colors', 'Memory', 'Storage', 'Camera', 'Rating',
       'Selling Price', 'Original Price', 'Mobile', 'Discount',
       'discount percentage'],
      dtype='object')

In [4]:
df.dtypes

Brands                  object
Models                  object
Colors                  object
Memory                  object
Storage                 object
Camera                  object
Rating                 float64
Selling Price            int64
Original Price           int64
Mobile                  object
Discount                 int64
discount percentage    float64
dtype: object

In [5]:
df.isna().any()

Brands                 False
Models                 False
Colors                 False
Memory                  True
Storage                 True
Camera                 False
Rating                  True
Selling Price          False
Original Price         False
Mobile                 False
Discount               False
discount percentage    False
dtype: bool

We will not investigate on duplicates since there is no way to identify single transactions with this kind of dataset.

## Adjusting the dataset 
First of all, we can notice that there is one Extra Variable that was not included in the description, which is 'Camera'.

In [6]:
df['Camera'].nunique()

1

In [7]:
df['Camera'].isna().any()

False

Since it only has one value, and there are no None values, then **this Variable is essentially useless** as it is a characteristic shared by all mobile phones present in the dataset, **hence we can drop it**.

In [8]:
df.drop('Camera',axis=1,inplace=True)

### Variable Analysis
- **Brands, Models, Colors**: Must be a categorical variables
- **Memory,Storage, Rating, Selling Price, Original Price, Discount, discount percentage**: Must be numerical variables.
<br> <br>
- **Brands**: Should be a Categorical Variable 

In [9]:
df['Brands'].unique()

array(['SAMSUNG', 'Nokia', 'realme', 'Infinix', 'Apple', 'GIONEE',
       'Xiaomi', 'OPPO', 'vivo', 'Lenovo', 'ASUS', 'Motorola', 'HTC',
       'Google Pixel', 'LG', 'POCO', 'IQOO'], dtype=object)

In [10]:
df['Brands']=df['Brands'].astype('category')

- **Models**: Should be a Categorical Variable 

In [11]:
df['Models'].unique()

array(['GALAXY M31S ', '3.2', 'C2 ', 'Note 5', 'iPhone 11 ', 'L800',
       'iPhone 13 Mini ', 'iPhone XR ', 'Galaxy E5 ', 'Redmi Note 11T 5G',
       'Galaxy A12 ', 'X7 5G ', 'A33', 'V23 5G', 'Y20A 2021 ',
       'Galaxy J2 Core ', 'K10 Plus ', 'Zenfone Max ', 'M', '1',
       'Pioneer P4S ', 'X3 ', 'Narzo 30A ', '8 Pro ', 'F205 Pro ',
       'A6600d40 ', 'Galaxy S9 Plus ', 'Galaxy A30s ', 'Galaxy J6 ',
       'Galaxy A03 Core', 'Galaxy Note 9 ', 'Hot 6 Pro',
       'Galaxy J2 - 2016 ', 'Metro 313', 'Desire 700', 'X3 SuperZoom ',
       '9 Pro+ 5G', 'Narzo 10 ', 'E5', '3', 'Zenfone Go 5.0 ',
       'Galaxy A22 5G ', 'Marathon M5 Plus ', 'GT Neo 2',
       'Galaxy A9 Pro ', 'Smart 7', '9i', 'iPhone 7 Plus ', 'iPhone 8 ',
       'Z Play', 'K4 Note ', 'Reno4 Pro Special Edition', 'Reno7 5G',
       '7i ', 'A390 ', 'Galaxy S6 Edge+ ', 'Stylus 3 ', 'Narzo 20A ',
       'Marathon ', 'Mi 11X Pro 5G ', 'G5 ', '6 Pro ', 'REDMI Note 9 ',
       'A5s', 'F1', 'One E9+', 'Galaxy M30S ', 'MI 11X 5G

In [12]:
df['Models']=df['Models'].astype('category')

- *Colors*

In [13]:
df['Colors'].unique()

array(['Mirage Black', 'Steel', 'Diamond Black', 'Ice Blue', 'Black',
       'Pink', 'Coral', 'Brown', 'Aquamarine Blue', 'Red', 'Nebula',
       'Moonlight Black', 'Sunshine Gold', 'Nebula Blue', 'Gold',
       'Sprite', 'Blue', 'Grey', 'White', 'Arctic White', 'Laser Blue',
       'Illuminating Yellow', 'Polaris Blue', 'Prism Crush Black',
       'Midnight Black', 'Bordeaux Red ', 'Green', 'That Green',
       'Fine Gold', 'Just Black', 'Violet', 'CHAMPAGNE GOLD', 'Neo Blue',
       'Purple', 'Prism Black', 'Jet Black', 'Silver', 'Yellow',
       'Starry Night', 'Startrails Blue', 'Ocean Blue', 'Fusion Blue',
       'Gold Platinum', 'Titan', 'Victory Blue', 'Celestial Silver',
       'Lightning Red', 'Pebble Grey', 'Gold Sepia', 'COSMIC BLACK',
       'Prism White', 'Glacier Green', 'Space Black', 'Champagne',
       'Stealth Black', 'Sky Blue', 'Forest Green', 'Black Sapphire',
       'Blazing Blue', 'Chroma White', 'Lake Green', 'Rust Red',
       'Crystal Silver', 'Lunar White', '

There are a lot of different colors!
We are, hence, firstly going to transform these fancy colors in their simplest forms _e.g. red, orange_

In [14]:
simple_colors=['red','blue','yellow','green','orange','purple','brown','pink','black','gray','white']
def find_simple_color(color):
    color = color.lower().replace(' ', '')
    for simple_color in simple_colors:
        if simple_color in color:
            return simple_color
    return color

In [15]:
new_colors = df['Colors'].apply(find_simple_color)

Now, let's check the remaining colors that were not caught by our previous transformation

In [16]:
list(set(new_colors).symmetric_difference(set(simple_colors)))

['starryglow',
 'titanium',
 'titansilver',
 'garlic',
 'goldandsilver',
 'starlight',
 'hazecrushsilver',
 'bronzegold',
 'frostedsilver',
 'crystalsymphony',
 'orchidgrey',
 'bladesilver',
 'lattegold',
 'rainbowsilver',
 'midnightjazz',
 'sunsetmelody',
 'carbongrey',
 'sapphiregradient',
 'illusionsky',
 'champagnegold',
 'glowinggalaxy',
 'glaringgold',
 'waterygrey',
 'sunsetdazzle',
 'goldsand',
 'starrynight',
 'meteorgrey',
 'richcranberry',
 'silver',
 'cyan',
 'quantumsilver',
 'grey/silver',
 'classicsilver',
 'fantasticrainbow',
 'lavenderviolet',
 'titan',
 'baltic',
 'tuscanycoral',
 'silverdiamond',
 'phantomviolet',
 'mint',
 'lavender',
 'neonspark',
 'steel',
 'ambergold',
 'frostedchampagne',
 'violet',
 'copper',
 'tradewgrey',
 'silvertitan',
 'silverwave',
 'royalgold',
 'frostedpearl',
 'frostedemerald',
 'graphite',
 'carbon',
 'rainbowfantasy',
 'moonlightjade',
 'quetzalcyan',
 'milkywaygrey',
 'darknebula',
 'night',
 'grey',
 'metalliccopper',
 'glorysilver

Here, we have to either add new colors, or to associate colors to simple ones. <br>
In this case, we decided to add only the gold color, that really isn't similar to any of the colors present in the original list

In [17]:
col_dict={'nordicsecret':'blue','electricgraphite':'black','diamondglow':'white','fantasticrainbow':'purple',
         'spacegrey':'gray','night':'black','richcranberry':'red','metalliccopper':'orange',
         'gunmetalgrey':'gray','alphagrey':'gray','smokysangria':'purple','darkgold':'gold',
          'cloudmint':'green','champagnegold':'gold','vanillamint':'blue','titaniumsapphire':'blue',
         'sunshinegold':'gold', 'glorysilver':'gray','ultraviolet':'purple','bronzegradient':'brown',
          'magicgold':'gold','silverdiamond':'gray', 'nebula':'purple','dazzlinggold':'gold','rainbowsilver':'gray',
         'lunargrey':'gray', 'tradewgrey':'gray', 'fierygold':'gold', 'concrete':'gray', 'mochagold':'gold',
         'spacesilver':'gray','racingsilver':'gray','gunmetalsilver':'gray','steel':'gray','starlight':'gold',
         'illusionsky':'blue', 'lavenderviolet':'purple','titan':'gray','meteorsilver':'gray','sunrisegold':'gold',
          'pastelsky':'blue', 'lunarsilver':'gray','amazingsilver':'gray','silverwave':'gray','topazgold':'gold','electricviolet':'purple',
          'mattegold':'gold', 'sunsetdazzle':'gold', 'cyan':'green','cybersilver':'gray','rosegold':'gold','flowingsilver':'gray',
          'carbon':'black','mint':'green','hazecrushsilver':'gray','symphonycyan':'green','iron':'gray','serenegold':'gold',
          'sunkissedleather':'gold', 'sunsetjazz':'gold','bladesilver':'gray', 'polarnight':'black',
         'powersilver':'gray','fjord':'green','aurorasilver':'gray','cloudnavy':'blue', 'prismcrushviolet':'purple','platinumgrey':'gray','darkgrey':'gray','moonlightsilver':'gray','gold':'gold','lasergrey':'gray','meteorgrey':'gray','copper':'orange','platinum':'gray',
          'graphite':'black','prismmagic':'purple','polishedgraphite':'black','prismcrushsilver':'gray','darknight':'black','starryglow':'gold','volcanicgrey':'gray',
          'carbongrey':'gray','celestialsnow':'blue','metallicgrey':'gray','goldsepia':'gold','quantumsilver':'gray','cream':'yellow','goldandsilver':'gray','verysilver':'gray','champagne':'yellow',
          'grey':'gray','twilightgrey':'gray','goldplatinum':'gold','silver':'gray','charcoal':'black','bronzegold':'brown','neonspark':'green','onion':'purple','titansilver':'gray',
          'lattegold':'gold','classicsilver':'gray','crystalsymphony':'gray','frostedemerald':'green','glowinggalaxy':'blue','coolgrey':'gray','darknebula':'purple','auraglow':'purple','mercurysilver':'gray','goldsand':'gold',
          'angelgold':'gold','ambergold':'gold','violet':'purple','maplegold':'gold','diamondruby':'red','tuscanycoral':'red','mysticbronze':'orange','metallicsilver':'gray','metallicsage':'green',
          'orchidgrey':'gray','shadowgrey':'gray','waterfallgrey':'gray','grey/silver':'gray','finegold':'gold','cyberteal':'green','metalgrey':'gray','oceanwave':'blue','moonlightjade':'green','saffrongrey':'gray',
          'polishedcopper':'brown','heartofocean':'blue','garlic':'red','blushgold':'gold','mithrilgrey':'gray','silvertitan':'gray','pacificpearl':'pink','starrynight':'blue','charcoalgrey':'gray','quetzalcyan':'green','sharkgrey':'gray',
          'kindofgrey':'gray','sprite':'green','azureglow':'blue','midnightjazz':'black','volcanogrey':'gray','sunriseflare':'gold','sunsetmelody':'gold','diamondsapphire':'blue','voyagergrey':'gray','mintcream':'green',
          'diamondflare':'purple','glaringgold':'gold','silvertitanium':'gray','baltic':'green','sapphiregradient':'blue','majesticgold':'gold','irischarcoal':'black','celestialsilver':'gray','phantomsilver':'gray','crystalsilver':'gray','rainbowfantasy':'purple',
          'middaydream':'purple','midnight':'black','auroragrey':'gray','milkywaygrey':'gray','auroradawn':'purple','midnightgrey':'gray','sapphirecyan':'green','awesomeviolet':'purple','coppergold':'brown','matteaqua':'blue',
          'phantomviolet':'purple','frostedgold':'gold','ice':'blue','royalgold':'gold','diamonddazzle':'white','coral':'red','titanium':'gray','lavender':'purple','polargold':'gold','pebblegrey':'gray',
          'glowinggold':'gold','frostedsilver':'gray','dusk':'blue','sand':'yellow','awesomemint':'green','darkpearl':'black','pacificsunrise':'yellow','frostedchampagne':'gray','moonlightgold':'gold',
          'frostedpearl':'white','cloudlavender':'purple','celestialmagic':'purple','aurora':'purple','waterygrey':'gray','richgrey':'gray' }


In [18]:
newcol2=new_colors.map(col_dict).fillna(new_colors)

In [19]:
##Let's check that we only have the original colors plus gold (hence the difference should yield only gold)
list(set(newcol2).symmetric_difference(set(simple_colors)))

['gold']

This is our new Colors column!

In [20]:
df['Colors']=newcol2.astype('category')

- *Memory* & *Storage* <br>
Memory is referring to the RAM, Random Access Memory, so the power of a device to use multiple apps simultaneously, while Storage is just the capacity to keep stuff saved inside the phone.
Both of these variables should be a Discrete Random Variable; hence, we must transform them. <br><br>

First, we have to find a uniform way to describe them, meaning that we shall find the most common, smallest unit of measure, starting from Gigabytes. <br>
For *Memory (RAM)*:


In [21]:
a=df[~df['Memory'].isna()]['Memory']
a[~a.str.endswith('GB')]

5         8 MB
36        2 MB
60      512 MB
116     512 MB
136     768 MB
         ...  
3079      4 MB
3092     64 MB
3096      4 MB
3107    512 MB
3108      4 MB
Name: Memory, Length: 164, dtype: object

In [22]:
b=a[~a.str.endswith('GB')]
b[~a.str.endswith('MB')]

Series([], Name: Memory, dtype: object)

For RAM the most atomic unit of measure is the **MB**! <br>
Let's now check the same for Storage.

In [23]:
a=df[~df['Storage'].isna()]['Storage']
a[~a.str.endswith('GB')]

5        16 MB
36        2 MB
120       2 MB
183       2 MB
197      16 MB
         ...  
3027      2 MB
3079      4 MB
3092    128 MB
3096      2 MB
3108      2 MB
Name: Storage, Length: 135, dtype: object

In [24]:
b=a[~a.str.endswith('GB')]
b[~a.str.endswith('MB')]

323     1 TB
591     1 TB
729     1 TB
799     1 TB
813     1 TB
831     1 TB
833     1 TB
903     1 TB
1185    1 TB
1285    1 TB
1903    1 TB
1974    1 TB
2032    1 TB
2932    1 TB
Name: Storage, dtype: object

It's Still the **MB**. <br>
We can now proceed with the transformation.

In [25]:
def bytes_to_num(s):
    if type(s)!=str:
        return 0
    if 'Expandable Upto' in s:
        s=s.replace('Expandable Upto','')
    if 'MB' in s:
        s=s.replace('MB','')
        if '.' in s:
            s=round(float(s),0)
        return int(s)
    elif 'GB' in s:
        s=s.replace('GB','')
        if '.' in s:
            s=round(float(s),0)
        return int(s)*1024
    elif 'TB' in s:
        s=s.replace('TB','')
        if '.' in s:
            s=round(float(s),0)
        return int(s)*1024*1024

In [26]:
df['Memory']=df['Memory'].apply(bytes_to_num)
df['Storage']=df['Storage'].apply(bytes_to_num)

- **Rating** <br>
Ratings are fine this way, only some None values but conceptually they should be, as they are, a *Continuos Random Variable*. 
<br>
<br>
- **Selling Price, Original Price and Discount**
In this case, they should be considered as a continuos random variable, but because Indian Paise are very small, usually commercial prices are *.00*, so it makes sense to consider them as integers.
<br>
<br>
- **Discount Percentage** <br>
At last, the discount percentage is fine, as it is a Continuos Random Variable and it is hence recognised as a float

In [27]:
df.dtypes

Brands                 category
Models                 category
Colors                 category
Memory                    int64
Storage                   int64
Rating                  float64
Selling Price             int64
Original Price            int64
Mobile                   object
Discount                  int64
discount percentage     float64
dtype: object

 - **Mobile**: 

In [28]:
df['Mobile'].isna().any()

False

In [29]:
df[['Brands','Models','Mobile']]

Brands          Models                Mobile
0     SAMSUNG    GALAXY M31S   SAMSUNG GALAXY M31S 
1       Nokia             3.2             Nokia 3.2
2      realme             C2             realme C2 
3     Infinix          Note 5        Infinix Note 5
4       Apple      iPhone 11       Apple iPhone 11 
...       ...             ...                   ...
3109     POCO       M4 Pro 5G        POCO M4 Pro 5G
3110    Nokia             225             Nokia 225
3111    Apple      iPhone SE       Apple iPhone SE 
3112    Apple  iPhone 13 Pro   Apple iPhone 13 Pro 
3113   GIONEE             F9             GIONEE F9 

[3114 rows x 3 columns]

As we can see here, **Mobile is just the string sum of Brand and Model**: it is **redundant**, and therefore we can drop it.

In [30]:
df.drop('Mobile', axis=1, inplace=True)

One other Boolean Variable that might be interesting for our analysis, as we can see in row 3112 is the presesence of **'Pro' models**.<br>
It is a naming strategy that companies use to differentiate the most high end version of a smartphone

In [31]:
def pro(s):
    if 'Pro' in s:
        return True
    elif 'PRO' in s:
        return True
    else:
        return False

In [32]:
df['pro']=df['Models'].apply(pro)

In [33]:
df.dtypes

Brands                 category
Models                 category
Colors                 category
Memory                    int64
Storage                   int64
Rating                  float64
Selling Price             int64
Original Price            int64
Discount                  int64
discount percentage     float64
pro                        bool
dtype: object

In [34]:
df.head()

Brands        Models Colors  Memory  Storage  Rating  Selling Price  \
0  SAMSUNG  GALAXY M31S   black    8192   131072     4.3          19330   
1    Nokia           3.2   gray    2048    16384     3.8          10199   
2   realme           C2   black    2048        0     4.4           6999   
3  Infinix        Note 5   blue    4096    65536     4.2          12999   
4    Apple    iPhone 11   black    4096    65536     4.6          49900   

   Original Price  Discount  discount percentage    pro  
0           20999      1669             7.947998  False  
1           10199         0             0.000000  False  
2            7999      1000            12.501563  False  
3           12999         0             0.000000  False  
4           49900         0             0.000000  False

After renaming the columns for simplicity matters, this is our Final Dataset:

In [35]:
df.rename(columns={'Brands':'brand','Models':'model','pro':'pro','Colors':'color','Memory':'memory',
                  'Storage':'storage','Rating':'rating','Selling Price':'selling_price',
                  'Original Price':'original_price','Discount':'discount',
                   'discount percentage':'disc_perc'}, inplace=True)

df=df[['brand','model','pro','color','memory','storage','rating','original_price','selling_price',
      'discount','disc_perc']]
df.to_csv('Clean_sales.csv', index=False)